mono/sgen/sgen-marksweep.c

```diff
@@ -2339,7 +2357,31 @@ major_scan_card_table (CardTableScanType scan_type, ScanCopyContext ctx)
 		if (!has_references)
 			continue;
 
-		scan_card_table_for_block (block, scan_type, ctx);
+		if (scan_type == CARDTABLE_SCAN_GLOBAL) {
+			gpointer *card_start = (gpointer*) sgen_card_table_get_card_scan_address ((mword)MS_BLOCK_FOR_BLOCK_INFO (block));
+			int i;
+			for (i = 0; i < CARDS_PER_BLOCK / sizeof(gpointer); i++) {
+				if (card_start [i]) {
```

Comment:
Please make this loop only check whether there's a card set.  All of the rest of the loop will only be executed once at most, so it shouldn't be in the loop.